cache: set default RDT and Block I/O classes when injecting container into cache.

klihub opened the pull request:
[Update: changed annotation syntax and this description accordingly.]

This PR implements #226 .

This patch set implements initializing the default RDT and Block I/O classes of containers during container creation/discovery. The classes are initialized to the first successfully found value from the following list:
  - container-specific class annotation value
  - pod-default class annotation value 
  - pod QoS class value

The class annotations are a more consistent variant of [these kubernetes conventions](https://kubesec.io/basics/metadata-annotations-seccomp-security-alpha-kubernetes-io-pod/), using the `rdtclass` and `blockioclass` keys in the usual resource-manager-specific `.cri-resource-manager.intel.com` key 'namespace'. Hence, the full key syntaxes are:
  - `rdtclass.cri-resource-manager.intel.com/container.$container`
  - `rdtclass.cri-resource-manager.intel.com/pod`
  - `blockioclass.cri-resource-manager.intel.com/container.$container`
  - `blockioclass.cri-resource-manager.intel.com/pod`


The patch series adds generic functions to the cache Pod and Container interfaces for getting the effective container-specific value of any annotation following the above described kubernetes-inspired conventions.

Therefore additionally,
  - `rdtclass.cri-resource-manager.intel.com` is treated as`rdtclass.cri-resource-manager.intel.com/pod`, and
  - `blockioclass.cri-resource-manager.intel.com` is treated as `blockioclass.cri-resource-manager.intel.com/pod`

Once this PR is merged, we'll file additional PRs for converting our other annotations to using this same convention and functions, whenever/wherever this make sense.

marquiz commented:
Quick comments, without looking at the code, yet.
> * Is CamelCase ok or should we stick to the all lowercase key variants `rdtclass` and `blockioclass` ?

I'm not a fan of camelcase here, quite the contrary. I'd also swap container<->class fields in the namespace, having:
- `container.rdtclass.cri-resource-manager.intel.com/$container`
- `rdtclass.cri-resource-manager.intel.com/pod`
- `container.blockioclass.cri-resource-manager.intel.com/$container`
- `blockioclass.cri-resource-manager.intel.com/pod`

I don't know if kubesec represents any established kubernetes convention(?) Just thinking that another naming convention for the containers-specifc annotations could be 
- `blockioclass.cri-resource-manager.intel.com/container.$container`
- `rdtclass.cri-resource-manager.intel.com/container.$container`


> * Regarding #232 I'm planning to get rid of the controller-level class mappings altogether. They are both a bit problematic and IMO after this patch set mostly unnecessary. Thoughts ?

👍 I have never liked this double-mapping. I think it probably just confuses users and makes the configuration harder.

klihub commented:
> Maybe the rdt/blockio class name mapping should already allow a shortcut, WDYT?
> 
> As a user, it would be pretty convenient to define cri-resmgr config:
> 
> ```
> blockio:
>   Classes:
>     LowPrioThrottled:
>       - Weight: 80
>     HighPrioFullSpeed:
>       - Weight: 400
>     Default:
>       - Weight: 100
> ```
> 
> And use these class names directly in annotations:
> 
> ```
> apiVersion: v1
> kind: Pod
>   name: bb-besteff
>   labels:
>     app: bb-besteff
>   annotations:
>     BlockIOClass.cri-resource-manager.intel.com/pod: HighPrioFullSpeed
> spec:
>   ...
> ```
> 
> Currently there's a problem: In `func (ctl *blockioctl) BlockIOClass(c cache.Container) string` the first line
> 
> ```
> cclass := c.GetBlockIOClass()
> ```
> 
> gets the direct block I/O class name (`HighPrioFullSpeed`). If this is not defined in the 2nd level block I/O class definitions in `resource-manager.blockio` classes (like "HighPrioFullSpeed": "HighPrioFullSpeed") but there is a default handler (like "*": "Default"), then the container gets assigned to the Default instead of the HighPrioFullSpeed class.

Sorry, I forgot to nudge you for comment #2 in the description for this PR (updated now): I would like to get rid of the controller-level mappings altogether, both for the RDT and Block I/O controllers. As it was used now, the main functionality was to assign the the QoS class as the default fallback class. If this PR gets merged that should not be necessary any more.

If after that there is still a need for mapping, then I would like to make that part of the rest of the Block I/O (and RDT) configuration, for instance by defining alias names for classes. But I would like to avoid that if we don't have a real use for it now.

klihub commented:
> Quick comments, without looking at the code, yet.
> 
> > * Is CamelCase ok or should we stick to the all lowercase key variants `rdtclass` and `blockioclass` ?
> 
> I'm not a fan of camelcase here, quite the contrary.

Ok, I'll switch to lowercase.

> I'd also swap container<->class fields in the namespace, having:
> 
> * `container.rdtclass.cri-resource-manager.intel.com/$container`
> * `rdtclass.cri-resource-manager.intel.com/pod`
> * `container.blockioclass.cri-resource-manager.intel.com/$container`
> * `blockioclass.cri-resource-manager.intel.com/pod`
> 
> I don't know if kubesec represents any established kubernetes convention(?) Just thinking that another naming convention for the containers-specifc annotations could be
 
> * `blockioclass.cri-resource-manager.intel.com/container.$container`
> * `rdtclass.cri-resource-manager.intel.com/container.$container`

There are other security and runtime related annotations with the same convention, IIRC. We once checked with Sasha, and that's why I ended up with this. I'm really not a big fan of the existing notation. My #1 hatred is that IMO it's stupid/confusing to have both DNS subdomain-like and /separated namespacing for keys. And if there was no prior art, I'd fully agree with you on the notation. But since there is prior art, why to make this more difficult for the user than it has to be by coming up with a new different convention ?

klihub commented on pkg/blockio/blockio.go:
```diff
@@ -73,6 +83,17 @@ var staticOciBlockIO = map[string]cgroups.OciBlockIOParameters{}
 // {"/dev/sda": "bfq"}
 var currentIOSchedulers map[string]string
 
```
Thanks, fixed.

askervin commented:
+1 for lowercase, +1 for direct class name mapping (and getting rid of current double mapping).

Perhaps we could enable direct mapping already here (in `BlockIOClass()` and in `RDTClass()`), so that user-defined mapping default ("*") would not change the class name if it was already referring to existing `blockio` or `rdt` class. Having it done, dropping the double mapping would not change annotation behavior later on.

klihub commented:
> +1 for lowercase, +1 for direct class name mapping (and getting rid of current double mapping).
> 
> Perhaps we could enable direct mapping already here (in `BlockIOClass()` and in `RDTClass()`), so that user-defined mapping default ("*") would not change the class name if it was already referring to existing `blockio` or `rdt` class. Having it done, dropping the double mapping would not change annotation behavior later on.

Hmm... I'm not entirely sure what do you mean by 'direct mapping' ? My plan was to simply have the next PR with two commits, one for removing each of the controller level class maps. After that there would be no mapping of any kind. Any class set for the container would need to be really defined for it to have any effect. There would be not be any builtin default/fallback class handling logic if a class does not exist (it would just result in an error).

marquiz commented:
> I'm not entirely sure what do you mean by 'direct mapping' ?

At least I understand it just like you describe it above 👍

askervin commented:
> Hmm... I'm not entirely sure what do you mean by 'direct mapping' ?

I meant using user-defined class names in annotations already after this patch.

> My plan was to simply have the next PR with two commits, one for removing each of the controller level class maps. After that there would be no mapping of any kind.

Makes sense.  And my concern about backwards compatibility does not make that much sense, because this sounds like we won't be making a release before removing the mapping. :)

> Any class set for the container would need to be really defined for it to have any effect. There would be not be any builtin default/fallback class handling logic if a class does not exist (it would just result in an error).

Giving up the default sounds ok. However, that made me wonder if we would still like to have something more powerful than pod-level control. Like

```
kubectl annotate namespace backup blockioclass.cri-resource-manager.intel.com=LowPriority
```

(On the other hand having annotations in one level only means less complexity for the code and debugging users issues...)

klihub commented:
> 
> > Any class set for the container would need to be really defined for it to have any effect. There would be not be any builtin default/fallback class handling logic if a class does not exist (it would just result in an error).
> 
> Giving up the default sounds ok.

I want to give up handling defaults *in the controller*. Having default/fallback handling logic at the blockio/rdt enforcement level is fine. And it might turn out to be necessary.

> However, that made me wonder if we would still like to have something more powerful than pod-level control. Like
> 
> ```
> kubectl annotate namespace backup blockioclass.cri-resource-manager.intel.com=LowPriority
> ```
> 
> (On the other hand having annotations in one level only means less complexity for the code and debugging users issues...)

Do you mean that this would be a way to annotate a global default for all pods/containers in the namespace `backup` ? It's a nice and compelling idea, especially if annotation on the namespace were inherited by all objects in the namespace. If they are not then we'd need to do some extra lifting via the agent to fetch those (and also probably watch for changes).

klihub commented:
> * `container.rdtclass.cri-resource-manager.intel.com/$container`
> * `rdtclass.cri-resource-manager.intel.com/pod`
> * `container.blockioclass.cri-resource-manager.intel.com/$container`
> * `blockioclass.cri-resource-manager.intel.com/pod`

Ok, so this is what we have now with the updated PR.

> I don't know if kubesec represents any established kubernetes convention(?) Just thinking that another naming convention for the containers-specifc annotations could be
> 
> * `blockioclass.cri-resource-manager.intel.com/container.$container`
> * `rdtclass.cri-resource-manager.intel.com/container.$container`

This would indeed feel more regular to me as well. Any preferences @kad for the annotation key syntax ?

askervin commented:
> Do you mean that this would be a way to annotate a global default for all pods/containers in the namespace `backup` ? It's a nice and compelling idea, especially if annotation on the namespace were inherited by all objects in the namespace. If they are not then we'd need to do some extra lifting via the agent to fetch those (and also probably watch for changes).

Yep, extra lifting would be needed to implement it this way.

There seem to be [PodPreset](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/) that enables injecting stuff to Pods at creation time. It's not the same as watched namespace annotations, but it may enable creation-time pod annotating with a wide ranger of match criteria.

> Then again, we have the configmap to configure (almost) everything (except the node config grouping) related to CRI resource manager, and such defaults could be also part of that configuration. With the current setup it certainly would be easier that way.

When I'm thinking  the configmap route further, it takes me pretty much back to various mappings - but those we're just about to get rid of. :)

klihub commented:
> > Do you mean that this would be a way to annotate a global default for all pods/containers in the namespace `backup` ? It's a nice and compelling idea, especially if annotation on the namespace were inherited by all objects in the namespace. If they are not then we'd need to do some extra lifting via the agent to fetch those (and also probably watch for changes).
> 
> Yep, extra lifting would be needed to implement it this way.
> 
> There seem to be [PodPreset](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/) that enables injecting stuff to Pods at creation time. It's not the same as watched namespace annotations, but it may enable creation-time pod annotating with a wide ranger of match criteria.

TBH, I'm not sure if that would help or hurt from the usability point of view. There would be one more unrelated component to configure to get what you want. That might be ok for users who already use `PodPreset`. For the rest it would shove `PodPreset` down their throats. 

> 
> > Then again, we have the configmap to configure (almost) everything (except the node config grouping) related to CRI resource manager, and such defaults could be also part of that configuration. With the current setup it certainly would be easier that way.
> 
> When I'm thinking the configmap route further, it takes me pretty much back to various mappings - but those we're just about to get rid of. :)

Well, you have to configure `cri-resmgr` for/in your cluster if you want to use it. Also, if you choose to exercise control over `RDT` or `Block I/O` you will anyway create the necessary configuration blocks for them. Coming from this angle, I was suggesting is that if we choose to provide a fallback then let's have a dedicated class name (preferably the rather intuitive `default`), which will be applied to containers that are not annotated to use any other class.

klihub commented:
@askervin Are you ok with merging this, or is there something to be fixed/addressed ?
@marquiz Could you ACK this is @askervin if/once is okay with tthis ?

marquiz commented on pkg/cri/resource-manager/cache/container.go:
```diff
@@ -111,6 +111,18 @@ func (c *container) fromCreateRequest(req *cri.CreateContainerRequest) error {
 
 	c.TopologyHints = topology.MergeTopologyHints(c.TopologyHints, getKubeletHint(c.GetCpusetCpus(), c.GetCpusetMems()))
 
+	class, ok := c.GetEffectiveAnnotation(RDTClassKey)
+	if !ok {
+		class = string(c.GetQOSClass())
+	}
+	c.SetRDTClass(class)
```
Just a thought: this functionality is duplicated in `fromListResponse(...)` so it could be split out into a separate method `setInitialRDTClass()`. And same thing for the blockio class

klihub replied:
Updated accordingly.

klihub commented on pkg/cri/resource-manager/cache/container.go:
```diff
@@ -228,8 +228,13 @@ func keysInNamespace(m map[string]string, namespace string) []string {
+			if split[0] == namespace {
+				keys = append(keys, split[1])
+			} else if namespace != "" && strings.HasSuffix(split[0], "."+namespace) {
+				key := strings.TrimSuffix(split[0], "."+namespace) + "/" + split[1]
```
> I think this just demonstrates that the "de-facto naming standard" of annotations we follow is kind of inconsistent/flawed.

I still agree, as I already did/told before. Since others don't chime in, I've updated the syntax to be consistent with our expectations. We can bike-shed more about this later if necessary.

klihub replied:
> If prolly end in trouble if we need to parse annotations in the plain `cri-resource-manager.intel.com` namespace

I didn't quite catch this train of thoughts...

marquiz replied:
I don't have the code, anymore, but my thinking was that if you'd like to get key in the `cri-resource-manager.intel.com` namespace and you'd have an annotations like
- `rdtclass.cri-resource-manager.intel.com/pod=foo`
- `cri-resource-manager.intel.com/bar=baz`

you would have got `rdtclass/pod=foo` and `bar=baz` although you really might've wanted to only have the latter

kad commented:
@klihub Have you checked annotation conventions of AppArmor ? https://kubernetes.io/docs/tutorials/clusters/apparmor/#pod-annotation

klihub commented:
> @klihub Have you checked annotation conventions of AppArmor ? https://kubernetes.io/docs/tutorials/clusters/apparmor/#pod-annotation

Yes (now). That seems to be a subset of the [seccomp annotations](https://kubesec.io/basics/metadata-annotations-seccomp-security-alpha-kubernetes-io-pod/) which the original patchset used. Based on the review comments I switched to the current, more consistent version inspired by the seccomp annotation logic.

kad commented:
> Based on the review comments I switched to the current, more consistent version inspired by the seccomp annotation logic.

ok, fine then. seccomp seems to be listed in k8s core constants and apparmor not. thus, yes, let's use seccomp blueprint.

klihub commented:
> > Based on the review comments I switched to the current, more consistent version inspired by the seccomp annotation logic.
> 
> ok, fine then. seccomp seems to be listed in k8s core constants and apparmor not. thus, yes, let's use seccomp blueprint.

@kad Originally, we used to use a notation logically identical to seccomp. But we switched to a syntax which is inspired by seccomp but not exactly the same (arguably more consistent since the pre-slash part of the annotation key is always the same). Seccomp uses
  - key.domain.name/pod for pod-global/default annotation, and
  - container.key.domain.name/$container for container-specific annotation

We diverge in the latter case by using:
  - key.domain.name/container.$container
